Framework/Algorithms/src/AbsorptionCorrection.cpp

```diff
@@ -163,42 +183,42 @@ void AbsorptionCorrection::exec() {
     calculateDistances(det, L2s);
 
     // If an indirect instrument, see if there's an efixed in the parameter map
-    double lambda_f = m_lambdaFixed;
+    double lambdaFixed = m_lambdaFixed;
     if (m_emode == 2) {
       try {
         Parameter_sptr par = pmap.get(&det, "Efixed");
         if (par) {
-          Unit_const_sptr energy = UnitFactory::Instance().create("Energy");
-          double factor, power;
-          energy->quickConversion(*UnitFactory::Instance().create("Wavelength"),
-                                  factor, power);
-          lambda_f = factor * std::pow(par->value<double>(), power);
+          lambdaFixed = energyToWavelength(par->value<double>());
         }
       } catch (std::runtime_error &) { /* Throws if a DetectorGroup, use single
                                           provided value */
       }
     }
 
-    const auto lambdas = m_inputWS->points(i);
+    // calculate the absorption coefficient for fixed wavelength
+    const double linearCoefAbsFixed = -m_material.linearAbsorpCoef(lambdaFixed);
+    const auto wavelengths = m_inputWS->points(i);
+    // these need to have the minus sign applied still
+    const auto linearCoefAbs =
+        m_material.linearAbsorpCoef(wavelengths.cbegin(), wavelengths.cend());
+
     // Get a reference to the Y's in the output WS for storing the factors
     auto &Y = correctionFactors->mutableY(i);
 
     // Loop through the bins in the current spectrum every m_xStep
     for (int64_t j = 0; j < specSize; j = j + m_xStep) {
-      const double lambda = lambdas[j];
-      if (m_emode == 0) // Elastic
-      {
-        Y[j] = this->doIntegration(lambda, L2s);
-      } else if (m_emode == 1) // Direct
-      {
-        Y[j] = this->doIntegration(m_lambdaFixed, lambda, L2s);
-      } else if (m_emode == 2) // Indirect
-      {
-        Y[j] = this->doIntegration(lambda, lambda_f, L2s);
+      if (m_emode == 0) { // Elastic
```

Comment:
I think we have `enum`s for these, please check `DeltaEMode.h`.

Reply:
Maybe even try out a `switch` statement here?

Reply:
I was bothered by the `int`. Thanks for letting me know about the `enum`.